### DISCUSSION TITLE
Encounter HttpException 500 caused by stream_copy_to_stream() when uploading

### DISCUSSION THREAD

**Discussion by HenryWu-OwO** [LycheeOrg/Lychee#2540] [Bugs] [answered]
### Lychee version

v5.5.1

### Which PHP version are you using?

PHP 8.2

### Detailed description of the problem

Just updated from 5.3.0 to 5.5.1, and encounters a 500 error when uploading. (everything is fine in 5.3.0)
The error message reads: 
                                        "stream_copy_to_stream(): Argument #1 ($from) must be of type resource, null given
                                        App\Image\Files\NativeLocalFile::append:119 stream_copy_to_stream(): Argument #1 ($from) must be of 
                                        type resource, null given; caused by"
Add `LIVEWIRE_ENABLED=false` to my `.env` file fixes the issue ( same as mentioned in issue#2285)

### Steps to reproduce the issue

1.Upload image in any albums
2.See error

### Diagnostics [REQUIRED]

```
Diagnostics
    -----------
        Warning: Dropbox import not working. dropbox_key is empty.
        Info: Latest version of PHP is 8.3
    
    System Information
    ------------------
        Lychee Version (release):                5.5.1
        DB Version:                              5.5.1
        
        composer install:                        --no-dev
        APP_ENV:                                 production
        APP_DEBUG:                               false
        APP_URL:                                 set
        APP_DIR:                                 default
        LOG_VIEWER_ENABLED:                      true
        LIVEWIRE_ENABLED:                        true
        
        System:                                  WINNT
        PHP Version:                             8.2.20
        PHP User agent:                          Lychee/4 (https://lycheeorg.github.io/)
        Timezone:                                Asia/Shanghai
        Max uploaded file size:                  1024M
        Max post size:                           1024M
        Livewire chunk size:                     12.00 MB
        Max execution time:                      0
        MySQL Version:                           8.3.0
        
        exec() Available:                        yes
        Imagick Available:                       1
        Imagick Enabled:                         1
        Imagick Version:                         1808
        GD Version:                              bundled (2.1.0 compatible)
        Number of foreign key:                   12 found.

 Config Information
    ------------------
        version:                                 050501
        check_for_updates:                       0
        sorting_photos_col:                      taken_at
        sorting_photos_order:                    ASC
        sorting_albums_col:                      max_taken_at
        sorting_albums_order:                    ASC
        imagick:                                 1
        skip_duplicates:                         0
        small_max_width:                         0
        small_max_height:                        480
        medium_max_width:                        2700
        medium_max_height:                       1800
        lang:                                    zh_CN
        image_overlay_type:                      desc
        default_license:                         reserved
        compression_quality:                     90
        grants_full_photo_access:                1
        delete_imported:                         0
        mod_frame_enabled:                       1
        mod_frame_refresh:                       30
        thumb_2x:                                1
        small_2x:                                1
        medium_2x:                               1
        landing_page_enable:                     0
        force_migration_in_production:           0
        search_public:                           0
        hide_version_number:                     0
        SL_enable:                               0
        SL_for_admin:                            0
        recent_age:                              1
        grants_download:                         0
        photos_wraparound:                       1
        map_display:                             0
        zip64:                                   1
        map_display_public:                      0
        map_provider:                            Wikimedia
        force_32bit_ids:                         0
        map_include_subalbums:                   0
        update_check_every_days:                 3
        has_exiftool:                            0
        share_button_visible:                    0
        import_via_symlink:                      0
        has_ffmpeg:                              0
        location_decoding:                       1
        location_decoding_timeout:               30
        location_show:                           1
        location_show_public:                    1
        rss_enable:                              0
        rss_recent_days:                         7
        rss_max_items:                           100
        prefer_available_xmp_metadata:           0
        editor_enabled:                          1
        lossless_optimization:                   0
        swipe_tolerance_x:                       150
        swipe_tolerance_y:                       250
        local_takestamp_video_formats:           .avi|.mov
        log_max_num_line:                        1000
        unlock_password_photos_with_url_param:   0
        nsfw_visible:                            1
        nsfw_blur:                               0
        nsfw_warning:                            0
        nsfw_warning_admin:                      0
        nsfw_banner_override:                    
        map_display_direction:                   1
        album_subtitle_type:                     oldstyle
        upload_processing_limit:                 4
        new_photos_notification:                 0
        legacy_id_redirection:                   1
        zip_deflate_level:                       6
        default_album_protection:                1
        album_decoration:                        layers
        album_decoration_orientation:            row
        allow_username_change:                   1
        auto_fix_orientation:                    1
        use_job_queues:                          1
        random_album_id:                         starred
        use_last_modified_date_when_no_exif_date: 0
        layout:                                  justified
        date_format_photo_thumb:                 M j, Y, g:i:s A e
        date_format_photo_overlay:               M j, Y, g:i:s A e
        date_format_sidebar_uploaded:            M j, Y, g:i:s A e
        date_format_sidebar_taken_at:            M j, Y, g:i:s A e
        date_format_hero_min_max:                F Y
        date_format_hero_created_at:             M j, Y, g:i:s A T
        date_format_album_thumb:                 M Y
        upload_chunk_size:                       0
        nsfw_banner_blur_backdrop:               0
        search_pagination_limit:                 1000
        search_minimum_length_required:          4
        photo_layout_justified_row_height:       320
        photo_layout_masonry_column_width:       300
        photo_layout_grid_column_width:          250
        photo_layout_square_column_width:        200
        photo_layout_gap:                        12
        display_thumb_album_overlay:             always
        display_thumb_photo_overlay:             hover
        default_album_thumb_aspect_ratio:        1/1
        use_album_compact_header:                0
        login_button_position:                   left
        back_button_enabled:                     0
        back_button_text:                        Return to Home
        current_job_processing_visible:          1
        SA_random_thumbs:                        0
        job_processing_queue_visible:            1
        login_required:                          0
        enable_unsorted:                         1
        enable_starred:                          1
        enable_recent:                           1
        enable_on_this_day:                      1
        login_required_root_only:                1
    
```

### Browser & System [REQUIRED]

Tried on both Archlinux ( Linux Kernel 6.11.0 ) and WIndows 11
Chrome 128.0.6613.137

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by ildyria** [comment, marked as answer]
That is a weird one.

Do you have a test bench? I would like you to check something for me.
Fortunately it currently works when you disable Livewire (so old process), but I would be curious to have you do a quick test on the future version. :)

**Reply by HenryWu-OwO** [reply]
I'm not sure what do you mean by "a test bench" (do you mean a server used for testing?), but I would be happy to help you check things out.

**Reply by ildyria** [reply]
pretty much yes.
See if you can reproduce your bug.
You can either use directly the `v6` branch... or try this alpha release: https://github.com/LycheeOrg/Lychee/releases/tag/v6.0.0-alpha1

**Reply by HenryWu-OwO** [reply]
I've tried v6.0-alpha release, the bug does not exist.

**Reply by ildyria** [reply]
Great. Thank you :)

**Comment by opravil-jan** [comment]
No it does not. I tried this with the same bug here #2550 and I get the same problem

**Reply by ildyria** [reply]
> Add `LIVEWIRE_ENABLED=false` to my `.env` file fixes the issue ( same as mentioned in issue#2285)

**Reply by opravil-jan** [reply]
When I change it to LIVEWIRE_ENABLED=false

I get this:

![image](https://github.com/user-attachments/assets/011088c9-9e54-429a-abad-727864513231)

**Reply by ildyria** [reply]
Can you dump your diagnostics ? that sounds weird to me. :(

**Reply by opravil-jan** [reply]
Diagnostics
    -----------
    Warning: Dropbox import not working. dropbox_key is empty.
    Warning: zend.assertions is disabled although Lychee is in debug mode. For easier debugging code generation for assertions should be enabled.
    Info: Latest version of PHP is 8.3
    Warning: git (software) is not available.
    Info: Found 4 medium that could be generated.
         You can use `php artisan lychee:generate_thumbs medium 4` to generate them.

    System Information
    ------------------
    Lychee Version (tag):                    v5.5.1 (656ffae) -- 1 tags behind v6.0.0-alpha1 (-0.00012833911891571 years ago)
    DB Version:                              5.5.1
    
    composer install:                        --no-dev
    APP_ENV:                                 production
    APP_DEBUG:                               true
    APP_URL:                                 set
    APP_DIR:                                 default
    LOG_VIEWER_ENABLED:                      true
    LIVEWIRE_ENABLED:                        false
    
    System:                                  Linux
    PHP Version:                             8.2.20
    PHP User agent:                          Lychee/4 (https://lycheeorg.github.io/)
    Timezone:                                Europe/Prague
    Max uploaded file size:                  100M
    Max post size:                           100M
    Livewire chunk size:                     12.00 MB
    Max execution time:                      3600
    PostgreSQL Version:                      PostgreSQL 16.2 (Debian 16.2-1.pgdg120+2) on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
    
    exec() Available:                        yes
    Imagick Available:                       1
    Imagick Enabled:                         1
    Imagick Version:                         1691
    GD Version:                              2.3.3
    Number of foreign key:                   12 found.

    Config Information
    ------------------
    version:                                 050501
    check_for_updates:                       0
    sorting_photos_col:                      title
    sorting_photos_order:                    ASC
    sorting_albums_col:                      description
    sorting_albums_order:                    DESC
    imagick:                                 1
    skip_duplicates:                         0
    small_max_width:                         0
    small_max_height:                        360
    medium_max_width:                        1920
    medium_max_height:                       1080
    lang:                                    cz
    image_overlay_type:                      desc
    default_license:                         reserved
    compression_quality:                     90
    grants_full_photo_access:                1
    delete_imported:                         0
    mod_frame_enabled:                       1
    mod_frame_refresh:                       30
    thumb_2x:                                1
    small_2x:                                1
    medium_2x:                               1
    landing_page_enable:                     1
    force_migration_in_production:           0
    search_public:                           0
    hide_version_number:                     0
    SL_enable:                               1
    SL_for_admin:                            1
    recent_age:                              1
    grants_download:                         0
    photos_wraparound:                       1
    map_display:                             1
    zip64:                                   1
    map_display_public:                      1
    map_provider:                            Wikimedia
    force_32bit_ids:                         0
    map_include_subalbums:                   1
    update_check_every_days:                 3
    has_exiftool:                            1
    share_button_visible:                    1
    has_ffmpeg:                              1
    import_via_symlink:                      0
    location_decoding:                       1
    location_decoding_timeout:               30
    location_show:                           1
    location_show_public:                    1
    rss_enable:                              0
    rss_recent_days:                         7
    rss_max_items:                           100
    prefer_available_xmp_metadata:           0
    editor_enabled:                          1
    lossless_optimization:                   0
    swipe_tolerance_x:                       150
    swipe_tolerance_y:                       250
    local_takestamp_video_formats:           .avi|.mov
    log_max_num_line:                        1000
    unlock_password_photos_with_url_param:   0
    nsfw_visible:                            1
    nsfw_blur:                               0
    nsfw_warning:                            0
    nsfw_warning_admin:                      0
    nsfw_banner_override:                    
    map_display_direction:                   1
    album_subtitle_type:                     oldstyle
    upload_processing_limit:                 4
    new_photos_notification:                 0
    legacy_id_redirection:                   1
    zip_deflate_level:                       6
    default_album_protection:                1
    album_decoration:                        layers
    album_decoration_orientation:            row
    allow_username_change:                   1
    auto_fix_orientation:                    1
    use_job_queues:                          0
    random_album_id:                         starred
    use_last_modified_date_when_no_exif_date: 0
    layout:                                  square
    date_format_photo_thumb:                 M j, Y, g:i:s A e
    date_format_photo_overlay:               M j, Y, g:i:s A e
    date_format_sidebar_uploaded:            M j, Y, g:i:s A e
    date_format_sidebar_taken_at:            M j, Y, g:i:s A e
    date_format_hero_min_max:                F Y
    date_format_hero_created_at:             M j, Y, g:i:s A T
    date_format_album_thumb:                 M Y
    upload_chunk_size:                       0
    nsfw_banner_blur_backdrop:               0
    search_pagination_limit:                 1000
    search_minimum_length_required:          4
    photo_layout_justified_row_height:       320
    photo_layout_masonry_column_width:       300
    photo_layout_grid_column_width:          250
    photo_layout_square_column_width:        200
    photo_layout_gap:                        12
    display_thumb_album_overlay:             always
    display_thumb_photo_overlay:             hover
    default_album_thumb_aspect_ratio:        1/1
    use_album_compact_header:                0
    login_button_position:                   left
    back_button_enabled:                     0
    back_button_text:                        Return to Home
    current_job_processing_visible:          1
    SA_random_thumbs:                        0
    job_processing_queue_visible:            1
    login_required:                          0
    enable_unsorted:                         1
    enable_starred:                          1
    enable_recent:                           1
    enable_on_this_day:                      1
    login_required_root_only:                1****

**Reply by ildyria** [reply]
Are you using a docker install?
Did you change anything in your `.env` with regard to mapping of folders ?

**Reply by opravil-jan** [reply]
docker-compose.yaml
```
version: '3.8'

services:
  lychee:
    image: lycheeorg/lychee
    container_name: lychee
    volumes:
        - ../../../../storage/conf:/conf
        - ../../../../storage/uploads:/uploads
        - ../../../../storage/sym:/sym
        - ../../../../storage/logs:/logs
    environment:
      APP_NAME: Lychee
      APP_ENV: production
      APP_KEY: 
      APP_DEBUG: "true"
      # This MUST contain the host name up to the Top Level Domain (tld) e.g. .com, .org etc.
      APP_URL: "https://www.petr-bures.com"
      APP_FORCE_HTTPS: "yes"
      DEBUGBAR_ENABLED: "true"
      LIVEWIRE_ENABLED: "true"
      LOG_VIEWER_ENABLED: "true"
      LEGACY_V4_REDIRECT: "false"
      DB_OLD_LYCHEE_PREFIX: 
      DB_CONNECTION: pgsql
      DB_HOST: xxxx
      DB_PORT: 5432
      DB_DATABASE: xxxx
      DB_USERNAME: xxxx
      DB_PASSWORD: "xxxx"
      DB_LOG_SQL: "false"
      DB_LOG_SQL_EXPLAIN: "false" #only for MySQL
      DB_LIST_FOREIGN_KEYS: "false"
      TIMEZONE: Europe/Prague
      CACHE_DRIVER: redis
      SESSION_DRIVER: redis
      SESSION_LIFETIME: 120
      QUEUE_CONNECTION: sync
      SECURITY_HEADER_HSTS_ENABLE: "false"
      SECURITY_HEADER_CSP_CONNECT_SRC: 
      SECURITY_HEADER_SCRIPT_SRC_ALLOW: 
      REDIS_SCHEME: "tcp"
      REDIS_HOST: xxxx
      REDIS_PASSWORD: null
      REDIS_PORT: 6379
      TRUSTED_PROXIES: "*"
      SKIP_PERMISSIONS_CHECKS: "yes"
      STARTUP_DELAY: 10
      ADMIN_USER: bures.petr
      ADMIN_PASSWORD: xxxx
    restart: unless-stopped

```

**Reply by ildyria** [reply]
That is a very weird, one.

One of my suspect is access rights on the following folders:
```
storage/image-tmp
storage/livewire-tmp
storage/image-tmp
```

The picture is being placed there before it is processed and moved to the upload folders. :|

**Reply by opravil-jan** [reply]
The rights look like this:

```
root@9327f592c06f:/var/www/html/Lychee# ls -la storage/ 
total 6
drwxr-xr-x 1 www-data www-data 4 Sep 29 06:17 .
drwxr-xr-x 1 www-data www-data 5 Sep 29 06:17 ..
drwxr-xr-x 3 www-data www-data 4 Jul  5 17:24 app
drwxr-xr-x 3 www-data www-data 3 Jul  5 17:24 bootstrap
drwxr-xr-x 2 www-data www-data 3 Jul  5 17:24 debugbar
drwxr-xr-x 2 www-data www-data 3 Jul  5 17:24 extract-jobs
drwxr-xr-x 1 www-data www-data 3 Jul  5 17:24 framework
drwxrwsr-x 2 www-data www-data 3 Jul  5 17:24 image-jobs
drwxr-xr-x 2 www-data www-data 3 Jul  5 17:24 image-tmp
drwxrwsr-x 2 www-data www-data 3 Jul  5 17:24 livewire-tmp
lrwxrwxrwx 1 root     root     4 Sep 29 06:17 logs -> /sym

```

**Reply by opravil-jan** [reply]
The rights are set by docker image and not changed by me.

**Reply by ildyria** [reply]
yeah I figured that one.
Space limitation on your docker ? :(

Any chance you can have a look in the sym/logs folder for some errors.log and see what the complaint is about if there are more details :/

Alternatively, could you check the request in the network tab of your browser and look at the response of the 500 error? The should be a stack trace in it.

**Reply by opravil-jan** [reply]
Here is part of the error.log

```
[2024-09-30 01:38:15] production.ERROR: ::Safe\realpath:9 An error occured
App\Image\Files\NativeLocalFile::getRealPath:226 An error occured; caused by
App\Image\Files\NativeLocalFile::getRealPath:228 An error occured; caused by
App\Exceptions\ModelDBException::create:50 Updating sym link failed; caused by
  
[2024-09-30 04:31:55] production.ERROR: ::<unknown>:752 Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes)
  
[2024-09-30 08:25:10] production.ERROR: ::Safe\realpath:9 An error occured
App\Image\Files\NativeLocalFile::getRealPath:226 An error occured; caused by
App\Image\Files\NativeLocalFile::getRealPath:228 An error occured; caused by
App\Exceptions\ModelDBException::create:50 Updating sym link failed; caused by
  
[2024-09-30 08:25:43] production.ERROR: ::<unknown>:1609 Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes)
  
[2024-09-30 08:26:39] production.ERROR: ::<unknown>:56 Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes)
  
[2024-09-30 08:31:25] production.ERROR: ::Safe\realpath:9 An error occured
App\Image\Files\NativeLocalFile::getRealPath:226 An error occured; caused by
App\Image\Files\NativeLocalFile::getRealPath:228 An error occured; caused by
App\Exceptions\ModelDBException::create:50 Updating sym link failed; caused by
  
[2024-09-30 08:49:55] production.ERROR: App\Actions\HoneyPot\BasePipe::throwNotFound:42 The route dist/backgrounds/uvodni_stranka.jpg could not be found.
  
[2024-09-30 08:51:55] production.ERROR: Illuminate\Foundation\Bootstrap\HandleExceptions::Illuminate\Foundation\Bootstrap\{closure}:50 stream_copy_to_stream(): Read of 8192 bytes failed with errno=21 Is a directory
App\Image\Files\NativeLocalFile::append:119 stream_copy_to_stream(): Read of 8192 bytes failed with errno=21 Is a directory; caused by
App\Image\Files\NativeLocalFile::append:129 stream_copy_to_stream(): Read of 8192 bytes failed with errno=21 Is a directory; caused by

```

**Reply by opravil-jan** [reply]
Any help please?

**Reply by d7415** [reply]
Looks like your container is running out of memory.
